pkg/controller/rpaasinstance/rpaasinstance_controller.go

```diff
@@ -702,3 +733,100 @@ func (_ boolPtrTransformer) Transformer(t reflect.Type) func(reflect.Value, refl
 		return nil
 	}
 }
+
+func portBelongsTo(port extensionsv1alpha1.AllocatedPort, instance *extensionsv1alpha1.RpaasInstance) bool {
+	if instance == nil {
+		return false
+	}
+	return instance.UID == port.Owner.UID && port.Owner.Namespace == instance.Namespace && port.Owner.RpaasName == instance.Name
+}
+
+func (r *ReconcileRpaasInstance) reconcilePorts(ctx context.Context, instance *extensionsv1alpha1.RpaasInstance, portCount int) ([]int, error) {
+	allocation := extensionsv1alpha1.RpaasPortAllocation{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: defaultPortAllocationResource,
+		},
+	}
+	err := r.client.Get(ctx, types.NamespacedName{
+		Name: defaultPortAllocationResource,
+	}, &allocation)
+	if err != nil {
+		if !errors.IsNotFound(err) {
+			return nil, err
+		}
+		err = r.client.Create(ctx, &allocation)
+		if err != nil {
+			return nil, err
+		}
+	}
+	var newPorts []extensionsv1alpha1.AllocatedPort
+	var usedSet bitset.BitSet
+	var instancePorts []int
+
+	// Loop through all allocated ports and remove ports from removed Nginx
+	// resources or from resources that have AllocateContainerPorts==false.
+	for _, port := range allocation.Spec.Ports {
+		var rpaas extensionsv1alpha1.RpaasInstance
+		err = r.client.Get(ctx, types.NamespacedName{
+			Namespace: port.Owner.Namespace,
+			Name:      port.Owner.RpaasName,
+		}, &rpaas)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				continue
+			}
+			return nil, err
+		}
+		if portBelongsTo(port, instance) {
+			if !instance.Spec.AllocateContainerPorts {
+				continue
+			}
+			instancePorts = append(instancePorts, int(port.Port))
+		}
+		if portBelongsTo(port, &rpaas) {
+			usedSet.Set(uint(port.Port))
+			newPorts = append(newPorts, port)
+		}
+	}
+
+	// If we should allocate ports and none are allocated yet we have to look
+	// for available ports and allocate them.
+	if instance != nil && instance.Spec.AllocateContainerPorts {
+		portMin := config.Get().PortRangeMin
+		portMax := config.Get().PortRangeMax
+		for port := portMin; port <= portMax; port++ {
```

Comment:
I'm worried about this implementation, we always are trying to allocate the first available ports. It also may lead to weird behavior in certain situations. Just to illustrate, imagine a scenario where some instance "A" (using the allocated ports 20000, 20001 and 20002) was deleted and a new one "B" was created soon after. The new one will got the free ports 20000, 20001, 20002. It means that, in a short period of time, the instance "B" (in starting mode) will receive the traffic of "A" (in terminating mode). This is especially true while the A's load balancer is not removed... Depending of the cloud provider, it may take a long time to be done.

In order to avoid that (and possibly others) scenario, I think it's safer we just increment the number of allocated ports. So, instead of beginning on `portMin` value, we start to look to available ports before the last allocated one. When reaching the `portMax` value, we could back to `portMin` and look from him.

Another simpler suggestion is to do a random based choice to get a free port. Although, we may have the same problems in extreme situations.

Reply:
That makes sense, the traffic would fix itself in a short time but it should be easier and safer to just start looking for a new port from the largest allocated one as you suggested.